src/providers/pkcs11_provider/asym_sign.rs

```diff
@@ -70,18 +79,13 @@ impl Pkcs11Provider {
         match self.backend.sign_init(session.session_handle(), &mech, key) {
             Ok(_) => {
                 info!("Signing operation initialized.");
-
-                // Build a valid ASN.1 DigestInfo DER-encoded structure by appending the hash to a
-                // DigestAlgorithmIdentifier value representing the SHA256 OID with no parameters.
-                // The OID used is: "2.16.840.1.101.3.4.2.1".
-                // It would be better to use the DigestInfo structure defined in this file but the
-                // AlgorithmIdentifier structure does not currently support the simple SHA256 OID.
-                // See Devolutions/picky-rs#19
-                let mut digest_info = vec![
-                    0x30, 0x31, 0x30, 0x0D, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04,
-                    0x02, 0x01, 0x05, 0x00, 0x04, 0x20,
-                ];
-                digest_info.append(&mut hash);
+                let digest_info = DigestInfo {
+                    oid: AlgorithmIdentifier::new_sha(SHAVariant::SHA2_256),
+                    digest: hash.into(),
+                };
+                let digest_info = picky_asn1_der::to_vec(&digest_info)
+                    // should not fail - if it does, there's some error in our stack
+                    .or(Err(ResponseStatus::PsaErrorGenericError))?;
```

Comment:
I am wondering about the return status here, is it right to return `GenericError`? I am thinking on how this function (`to_vec`) could fail. Could that be because `digest` is not a `SHA256`, in which case this could be a `PsaErrorInvalidArgument`?

Reply:
I'll look into the code for `to_vec`, but I doubt it - I think it simply serializes the contents of that reference, regardless of what's in there.

Reply:
It only takes the value and serializes it, no checks done:

https://github.com/Devolutions/picky-rs/blob/master/picky-asn1-der/src/ser/mod.rs#L17

Reply:
Yes I think that is the most appropriate one then.